test/integration/lib/cluster_viz_service_test.rb

```diff
@@ -0,0 +1,271 @@
+
+    cluster = @study.default_cluster
+    metadata = @study.cell_metadata.by_name_and_type('species', 'group')
+    annotation = {name: metadata.name, type: metadata.annotation_type, scope: 'study'}
+    coordinates = ClusterVizService.load_cluster_group_data_array_points(@study, cluster, annotation)
+    assert_equal metadata.values.sort, coordinates.keys.sort
+    trace_attributes = [:x, :y, :z, :cells, :annotations, :name]
+    coordinates.each do |name, trace|
+      trace_attributes.each do |attribute|
+        assert trace.try(:[], attribute).present?, "Did not find any values for #{name}:#{attribute}"
+      end
+    end
+
+    puts "#{File.basename(__FILE__)}: #{self.method_name} successful!"
+  end
+
+  test 'should transform cluster coordinates for visualization' do
+    puts "#{File.basename(__FILE__)}: #{self.method_name}"
+
+    cluster = @study.default_cluster
+    metadata = @study.cell_metadata.by_name_and_type('disease', 'group')
+    annotation = {name: metadata.name, type: metadata.annotation_type, scope: 'study'}
+    coordinates = ClusterVizService.load_cluster_group_data_array_points(@study, cluster, annotation)
+    transformed_coords = ClusterVizService.transform_coordinates(coordinates, @study, cluster, annotation)
+    assert_equal coordinates.keys.size, transformed_coords.size,
+                 "Did not find correct number of traces; expected #{coordinates.keys.size} but found #{transformed_coords.size}"
+    expected_keys = [:x, :y, :cells, :text, :name, :type, :mode, :marker, :opacity, :annotations, :z, :textposition].sort
+    transformed_coords.each do |trace|
+      keys = trace.keys.sort
+      assert_equal expected_keys, keys
+      assert_equal 'scatter3d', trace[:type]
+      assert_equal 'markers', trace[:mode]
+    end
+
+    puts "#{File.basename(__FILE__)}: #{self.method_name} successful!"
+  end
+
+  test 'should load coordinate labels as annotations' do
+    puts "#{File.basename(__FILE__)}: #{self.method_name}"
+
+    cluster = @study.default_cluster
+    FactoryBot.create(:coordinate_label_file,
+                      cluster: cluster, name: 'coordinate_labels.txt', study: @study,
+                      label_input: {
+                          x: [1, 2, 3],
+                          y: [4, 5, 6],
+                          z: [7, 8, 9],
+                          text: ['Group 1', 'Group 2', 'Group 3']
+                      })
+    labels = ClusterVizService.load_cluster_group_coordinate_labels(cluster)
+    assert_equal 3, labels.size, "Did not find correct number of labels; expected 3 but found #{labels.size}"
+    selected_label = labels.first
+    assert_equal 1, selected_label[:x]
+    assert_equal 4, selected_label[:y]
+    assert_equal 7, selected_label[:z]
+    assert_equal 'Group 1', selected_label[:text]
+    assert selected_label[:font].present?
+    font_keys = [:family, :size, :color]
+    assert_equal font_keys, selected_label[:font].keys
```

Comment:
Tangentially: font strikes me as better handled in the front-end, by static assets.

Reply:
This is mainly validating the structure of the returned object, but point taken.